birdhouse/docker-compose.yml

```diff
@@ -409,6 +409,7 @@ services:
       JUPYTERHUB_IMAGE_SELECTION_NAMES: ${JUPYTERHUB_IMAGE_SELECTION_NAMES}
       DOCKER_NETWORK_NAME: jupyterhub_network
       JUPYTERHUB_USER_DATA_DIR: ${JUPYTERHUB_USER_DATA_DIR}
+      WORKSPACE_DIR: ${JUPYTERHUB_USER_DATA_DIR}
```

Comment:
I think the override strategy of this variable by Cowbird works at the moment only because components are *always* loaded after this "base" docker-compose. 

However, with eventual PR #296 that will make JupyterHub also a component, it could potentially break depending on load order. It think we should find a way to define some kind of `USER_WORKSPACE_DIR` variable that is resolved conditionally considering all loaded components, and the `jupyter` service would simply use it without need to override in `cowbird`'s docker-extra. So, in other words, the override should happen at the `pavics-compose.sh` or the JupyterHub config level (if not already the case as suggested by the `if os.environ['WORKSPACE_DIR'] != jupyterhub_data_dir:` check?) instead of `docker-compose` load order.

@mishaschwartz Tagging you as this item might have some considerations with your PR. Any ideas on how to best handle it so solutions align after both PRs are merged.

Reply:
I hate to sound like a broken record here but why not use parameter expansion?

```sh
# default.env
export DATA_PERSIST_ROOT="/data"
export JUPYTERHUB_USER_DATA_DIR='${DATA_PERSIST_ROOT}/${JUPYTER_HUB_USER_DATA_BASE_DIR:-jupyterhub_user_data}'

# components/cowbird/default.env
export JUPYTERHUB_USER_DATA_BASE_DIR="user_workspaces"
```

Also, maybe I'm missing something here but why does the volume mounted at `JUPYTERHUB_USER_DATA_DIR` need to be at the same path as on the host?

Reply:
It is possible that some `env.local` could already define `export JUPYTERHUB_USER_DATA_DIR` explicitly, so the required path to be adjusted by cowbird to make this feature work would never be applied. The `WORKSPACE_DIR` variable however is new and specific to this feature, so if it is overridden and things don't work, it is more understandable.

---
For volumes.
**TL;DR** : not required, but way easier / less error prone

They don't "need" (from a technical point) to be the same locations, but having different paths implies that Cowbird/Magpie services always have to perform some kind of reverse mapping of the location to generate the correct links that would be seen from inside the docker images when they mount these volumes. The `cowbird-worker` resolves paths/permissions across all users, so it needs the top `/data/user_workspaces` directory, JupyerHub, however, needs only the user-specific workspace. If we did the following:

```yaml
services:
  jupyter:
    volumes:
      - "/data/jupyterhub_user_data_dir/userX:/notebooks"
      - "/data/user_workspaces/userX:/workspace"
  cowbird-worker:
    volumes:
      - "/data/jupyterhub_user_data_dir:/notebooks"
      - "/data/user_workspaces:/user_workspaces"
```

Then a link `/user_workspaces/userX/notebooks -> /notebooks/userX` created from inside `cowbird-worker` will not work inside `jupyter`, since under its in-docker context, the target path would not exist (directly `/notebooks` is mounted, not `/notebooks/userX`). Cowbird would have to assume that `/notebooks` and `/workspace` will be the mount points in `jupyter`, to create `/user_workspaces/userX/notebooks -> /notebooks`, although `/notebooks` from Cowbird's point of view is an entirely different location. This can lead to a lot of confusion, and incorrect file permission configuration, when trying to manage which links to create/delete under the different `jupyter`/`cowbird` in-docker contexts.  

If instead of `/data/jupyterhub_user_data_dir:/notebooks`, we used let say `/data/jupyterhub_user_data_dir:/all-users-notebooks` to avoid the `/notebooks` location confusion, then Cowbird would instead have to "blindly" create `/user_workspaces/userX/notebooks -> /notebooks`, although for Cowbird, that target `/notebooks` wouldn't exist. All OS links/chmod/etc. calls would have to do a lot of error handling to ignore "missing" paths and force link creation.

Reply:
@mishaschwartz 
Since this PR has been ready to merge for a few days, we were wondering if you were OK with me merging it on `master` and you could check to integrate the changes in your current PR [#296](https://github.com/bird-house/birdhouse-deploy/pull/296)?

Would that be okay, or would you have another proposition? Thank you!

Reply:
@ChaamC That's fine with me. Once this is merged in to master I'll update my branch